### PR TITLE
[TT-11726]: This optimizes the deletion of keys using a single command

### PR DESCRIPTION
**Pull request by kofoworola**
### **User description**

[TT-11726](https://tyktech.atlassian.net/browse/TT-11726)

This is the benchmark test for addKey before the changes were made
Screenshot 2024-08-12 at 07 15 46

This is for after the changes
Screenshot 2024-08-12 at 07 17 17

There is a significant reduction in the time per iteration
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

[TT-11726]: https://tyktech.atlassian.net/browse/TT-11726?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

___

### **PR Type**
Enhancement

___

### **Description**
- Optimized the deletion of keys in `gateway/auth_manager.go` by using a single `DeleteKeys` command instead of multiple `DeleteRawKey` calls.
- Added logging to track deleted keys with allowance scopes.
- Improved code efficiency and readability by consolidating key deletions into a single operation.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>auth_manager.go</strong><dd><code>Optimize key deletion using a single command</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/auth_manager.go

<li>Optimized key deletion by using a single command.<br> <li> Replaced multiple <code>DeleteRawKey</code> calls with <code>DeleteKeys</code>.<br> <li> Added logging for deleted keys with scopes.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6427/files#diff-311e18b071d244ed1615c0019215b633278679df373288b3451bcc5cf7c52c4e">+12/-5</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: gateway/auth_manager.go [hunk 1]
```diff
@@ -74,12 +77,33 @@ func (b *DefaultSessionManager) ResetQuota(keyName string, session *user.Session
 
 	rateLimiterSentinelKey := RateLimitKeyPrefix + keyName + ".BLOCKED"
 
-	// Clear the rate limiter
-	b.store.DeleteRawKey(rateLimiterSentinelKey)
+	// Clear the rate limiter and
 	// Fix the raw key
-	b.store.DeleteRawKey(rawKey)
+	keys := make([]string, 0)
```
Comment (titpetric) [inline gateway/auth_manager.go:82]:
```suggestion
```

File: gateway/auth_manager.go [hunk 2]
```diff
@@ -74,12 +77,33 @@ func (b *DefaultSessionManager) ResetQuota(keyName string, session *user.Session
...
+	keys = append(keys, rateLimiterSentinelKey, rawKey)
```
Comment (titpetric) [inline gateway/auth_manager.go:83]:
```suggestion
	keys := []string{rateLimiterSentinelKey, rawKey}
```

File: gateway/auth_manager_test.go [hunk 3]
```diff
@@ -510,7 +510,7 @@ func (c *countingStorageHandler) GetKeysAndValuesWithFilter(s string) map[string
 	return nil
 }
 
-func (c *countingStorageHandler) DeleteKeys(strings []string) bool {
+func (c *countingStorageHandler) DeleteKeys(keys []string) bool {
```
Comment (titpetric) [inline gateway/auth_manager_test.go:513]:
```suggestion
func (c *countingStorageHandler) DeleteKeys(_ []string) bool {
```

File: gateway/auth_manager.go [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
 
 import (
+	"context"
 	"encoding/base64"
 	"encoding/json"
 	"strings"
 	"time"
 
+	"github.com/TykTechnologies/tyk/internal/redis"
+
```
Comment (titpetric) [inline gateway/auth_manager.go:11]: reorder the import together with other TykTechnologies imports (move down about 4 lines)

File: gateway/auth_manager.go [hunk 5]
```diff
@@ -74,12 +77,32 @@ func (b *DefaultSessionManager) ResetQuota(keyName string, session *user.Session
...
 	// Fix the raw key
-	b.store.DeleteRawKey(rawKey)
+	keys := []string{rateLimiterSentinelKey, rawKey}
+	for _, acl := range session.AccessRights {
+		if acl.AllowanceScope == "" {
+			continue
+		}
+		keys = append(keys, QuotaKeyPrefix+acl.AllowanceScope+"-"+keyName)
+	}
+	type clientProvider interface {
```
Comment (jeffy-mathew) [inline gateway/auth_manager.go:89]: we could use `b.store.DeleteKeys(keys)` to bulk delete keys

File: gateway/auth_manager.go [hunk 6]
```diff
@@ -74,12 +77,32 @@ func (b *DefaultSessionManager) ResetQuota(keyName string, session *user.Session
 
 	rateLimiterSentinelKey := RateLimitKeyPrefix + keyName + ".BLOCKED"
 
-	// Clear the rate limiter
-	b.store.DeleteRawKey(rateLimiterSentinelKey)
+	// Clear the rate limiter and
 	// Fix the raw key
-	b.store.DeleteRawKey(rawKey)
+	keys := []string{rateLimiterSentinelKey, rawKey}
+	for _, acl := range session.AccessRights {
```
Comment (jeffy-mathew) [inline gateway/auth_manager.go:83]: shall we wrap these into a function `rawKeysWithAllowanceScope` that returns the keys to be deleted with allowance scope?

File: gateway/api_test.go [hunk 7]
```diff
@@ -591,6 +591,61 @@ func TestKeyHandler_UpdateKey(t *testing.T) {
 	})
 }
 
+func BenchmarkKeyHandler_CreateKeyHandler(b *testing.B) {
+	ts := StartTest(func(globalConf *config.Config) {
```
Comment (jeffy-mathew) [inline gateway/api_test.go:595]:
```suggestion
	ts := StartTest(nil) {
```

File: gateway/api.go [hunk 8]
```diff
@@ -1957,7 +1957,7 @@ func (gw *Gateway) createKeyHandler(w http.ResponseWriter, r *http.Request) {
 			"status": "fail",
 			"err":    err,
 		}).Error("Key creation failed.")
-		doJSONWrite(w, http.StatusInternalServerError, apiError("Unmarshalling failed"))
+		doJSONWrite(w, http.StatusInternalServerError, apiError("Unmarshalling faileddd"))
```
Comment (jeffy-mathew) [inline gateway/api.go:1960]: typo in `faileddd`

File: gateway/auth_manager.go (hunk 6 above)
Comment (kofoworola) [reply gateway/auth_manager.go:83]: thanks for pointing this

File: gateway/auth_manager.go (hunk 5 above)
Comment (kofoworola) [reply gateway/auth_manager.go:82]: i tried this earlier, the deletekeys adds a prefix and does not delete the raw keys, and changing the method signature would apparently break plugins

File: gateway/rpc_storage_handler.go [hunk 9]
```diff
@@ -533,6 +536,26 @@ func (r *RPCStorageHandler) DeleteKey(keyName string) bool {
...
+		)
+
+		if r.IsRetriableError(err) {
+			if rpc.Login() {
+				return r.DeleteRawKeys(keys)
+			}
+		}
+	}
+
+	return ok == true
```
Comment (jeffy-mathew) [inline gateway/rpc_storage_handler.go:556]:
```suggestion
	return ok
```

**Review by jeffy-mathew — APPROVED** [review]

**Comment by titpetric** [comment]
@jeffy-mathew @kofoworola this adds new functions to storage, but they don't return errors, should have caught it

**Pull request merged**